### PR TITLE
[zk-sdk] Add error types and zero ciphertext instruction

### PR DESCRIPTION
**Pull request by samkim-crypto**
#### Problem
The proof data and instructions for the new elgamal program in zk-sdk are not added yet.

#### Summary of Changes
I added the zero-ciphertext instruction to start. Once this PR is merged after some feedback, then I think I can add the rest of the proof data in slightly larger chunk PRs.

- [7c0e380](https://github.com/anza-xyz/agave/pull/1453/commits/7c0e38048af5b1d8155439ce7c948e8bb06b1f42): I added the `ProofGenerationError` and `ProofVerificationError`. These types remain largely the same as in zk-token-sdk and in particular, wraps `RangeProofError` and `SigmaProofError`.
- [d701214](https://github.com/anza-xyz/agave/pull/1453/commits/d70121472302650038442e13e2a6ba6da54e2d7a): I just copied the `zero_balance` instruction data module from zk-token-sdk verbatim to make the diffs in the next commit easier to view.
- [8468880](https://github.com/anza-xyz/agave/pull/1453/commits/846888072a69292ece9c423fa4b252287ec50c37): I updated the `zero_balance` instruction data. The changes are essentially updating `ZeroBalance...` to `ZeroCiphertext...` and re-organizing the imports.
- [25c3c88](https://github.com/anza-xyz/agave/pull/1453/commits/25c3c88b9d88840831f50b1b8f135522635b74eb): Added the `VerifyZeroCiphertext` instruction in the elgamal proof program.

Fixes #

### CODE REVIEW

File: zk-sdk/src/elgamal_program/errors.rs [hunk 1]
```diff
@@ -1,7 +1,54 @@
...
+    thiserror::Error,
+};
+
+#[cfg(not(target_os = "solana"))]
+#[derive(Error, Clone, Debug, Eq, PartialEq)]
+pub enum ProofGenerationError {
+    #[error("not enough funds in account")]
+    NotEnoughFunds,
+    #[error("transfer fee calculation error")]
+    FeeCalculation,
```
Comment (joncinque) [inline zk-sdk/src/elgamal_program/errors.rs]: Should these errors be changed or removed?

**Review by joncinque — COMMENTED** [review]
Looks good overall, just the one comment about the error names

File: zk-sdk/src/elgamal_program/errors.rs (hunk 1 above)
Comment (samkim-crypto) [reply zk-sdk/src/elgamal_program/errors.rs]: Oh great point! I removed it. Thanks!

Comment (joncinque) [reply zk-sdk/src/elgamal_program/errors.rs]: Ok cool! Can you also get rid of `NotEnoughFunds`?

Comment (samkim-crypto) [reply zk-sdk/src/elgamal_program/errors.rs]: Oh oops, removed. Thanks!

**Review by joncinque — APPROVED** [review]
Looks great!

**Pull request merged**